### PR TITLE
[OptApp] Update Filters

### PR DESCRIPTION
**Pull request by sunethwarna**
**📝 Description**
This PR updates both implciit and explicit filters to support `Filter::Factory`. Following changes are made:

**Explicit Filtering**
1. There is no need to have seperate `DampingFunction` anymore because, I unified all the filter functions including `sigmoidal` function. Now all of them can be used either as filter or damping. Following figure shows the filter functions: (May be in future, we can have seperate settings for these functions, so we can tweak some of the coeffs used in these filter functions).
filters

**VertexMorphingShapeControl**
1. `bulk_surface_shape_control` is moved out from the `VertexMorphingShapeControl`. The reasoning is:, `bulk_surface_shape_control` shape control handles the mesh motion also in the filter. But this is not always required (such as in the shell surface optimization). The current `VertexMorphingShapeControl` always assumes it works only on the surface model parts, and if required it can use the mesh motion analysis to optimize the mesh in the case of solid geometry optimzation. I will create another shape control which inherits from the `VertexMorphingShapeControl` to use the `bulk_surface_shape_control`. The `bulk_surface_shape_control` will not be added to the FilterFactory because, it cannot be used on a generic field.
3. No unfilter method is used to get the `control_field`. The reasoning is, the explicit filters does not have `Unfilter` method, and it is not required to have the correct initial `control_field` as long as we do the update in the physical space correctly as shown below using the initial physical field where $`\underline{s}_k`$ is the $`k^{th}`$ iteration physical field and $`\underline{\Delta p}_i`$ is the $`i^{th}`$ iteration control space update and $`\mathbf{G}_i`$ is the $`i^{th}`$ iterations filtering matrix.

```math
\underline{s}_k = \underline{s}_0 + \sum_{i=1}^k \underline{\Delta s}_i = \underline{s}_0 + \sum_{i=1}^k \mathbf{G}_i\underline{\Delta p}_i
```

**ShellThicknessControl**
1. Shell thickness is also now filters only the $`\underline{\Delta p}_i`$ instead of the absolute control space values and correctly does the update in the physical space using $`\underline{s}_0`$ .
2. Initial control field for shell thickness also does not use the `Unfilter` method.
4. The updating is correctly done using the $`\underline{s}_0`$
5. Damping is applied as zero values in implcit becuase, now filtering is only done on $`\underline{\Delta p}_i`$, then damped model part regions should have $`\underline{\Delta s}_i = 0`$ which is the Dirichlet BC for system of equations. Advantage of this uniform damping in implicit case is, then user can have any arbitray function applied for the physical space variable in the damped region.

Followings is the comparison (only ran 100 iterations, not till the convergence):
implicit_explicit_filtering_comparison

:warning:  **Tests will be failing. I will fix the tests, once we agree on the changes** :warning: 

FYI: @talhah-ansari 

**🆕 Changelog**
- Implemented `ExplicitFilter`
- Implemneted `ImplicitFilter`
- Update `Helmholtz` solvers to support `ImplicitFilter`
- Added tests to `ExplicitFilter`

### CODE REVIEW

File: applications/OptimizationApplication/custom_utilities/filtering/explicit_filter_utils.cpp [hunk 1]
```diff
@@ -101,82 +80,75 @@ double GetDomainSize(
...
         const double DomainSize)
     {
-        return FilterFunctionWeight * DomainSize;
+        return Value / DomainSize;
     }
 };
 
-template <class TEntityType, class TWeightComputationType>
+template <class TEntityType>
 void ComputeWeightForAllNeighbors(
```
Comment (RezaNajian) [inline applications/OptimizationApplication/custom_utilities/filtering/explicit_filter_utils.cpp:104]: The implementation of this function is changed ! and it has an important impact on the performance. Please take the original implementation of the ComputeWeightForAllNeighbors from explicit_filter.cpp in the master branch.

Comment (RezaNajian) [reply applications/OptimizationApplication/custom_utilities/filtering/explicit_filter_utils.cpp:104]: The damping coefficients should be integrated into the function as it was before !

Comment (RezaNajian) [reply applications/OptimizationApplication/custom_utilities/filtering/explicit_filter_utils.cpp:104]: Please update forward/backward... as well. I highly recommend not changing the current implementation without providing the needs and reasons!

File: applications/OptimizationApplication/python_scripts/controls/shape/vertex_morphing_shape_control.py [hunk 2]
```diff
@@ -150,29 +36,16 @@ def __init__(self, name: str, model: Kratos.Model, parameters: Kratos.Parameters
...
-                "max_nodes_in_filter_radius": 1000,
-                "linear_solver_settings" : {}
-            },
-            "mesh_motion_settings" : {},
-            "output_all_fields": false,
-            "fixed_model_parts": {},
-            "utilities": []
+            "filter_settings"            : {},
+            "use_mesh_motion_solver"     : false,
+            "skin_model_part_names"      : [],
```
Comment (RezaNajian) [inline applications/OptimizationApplication/python_scripts/controls/shape/vertex_morphing_shape_control.py:41]: skin_model_part_names is not self explaining please change it to sth like fixed_model_parts and also get the components for fixing

File: applications/OptimizationApplication/python_scripts/controls/shape/vertex_morphing_shape_control.py [hunk 3]
```diff
@@ -181,93 +54,66 @@ def __init__(self, name: str, model: Kratos.Model, parameters: Kratos.Parameters
...
-            self.control_field = self.filter.UnFilterField(self.GetPhysicalField())
-            self._SetFixedModelPartValues(True)
+        self.filter.SetComponentDataView(ComponentDataView(self, self.optimization_problem))
+        self.filter.Initialize()
+
+        # getting UnfilteredField of the PhysicalField is not possible with
+        # explicit solvers, and is not useful for this control hence, this
+        # will only update the shape changes to the existing shape. Hence,
+        # taking the actual physical field as the control field initially.
+        self.control_field = self.GetPhysicalField()
```
Comment (RezaNajian) [inline applications/OptimizationApplication/python_scripts/controls/shape/vertex_morphing_shape_control.py:80]: that's not true ! Please do not change the original implementation without providing notices and reasons. Please take the original implementation which distinguishes between explicit and implicit filters. Implicit filter provides the control field, hence filters should provide unfilterfield, in case of explicit, zerofield should be returned

Comment (RezaNajian) [reply applications/OptimizationApplication/python_scripts/controls/shape/vertex_morphing_shape_control.py:80]: Please complete the comments with the fact that implicit filters provide the control field by unfiltering however explicit filters only return zero field since thy can not !

Comment (RezaNajian) [reply applications/OptimizationApplication/python_scripts/controls/shape/vertex_morphing_shape_control.py:80]: In overall, please only use the unfilterfield function of the filter to get the control field

File: applications/OptimizationApplication/python_scripts/controls/shape/vertex_morphing_shape_control.py [hunk 4]
```diff
@@ -181,93 +54,66 @@ def __init__(self, name: str, model: Kratos.Model, parameters: Kratos.Parameters
...
 
-        # initialize_mesh_motion
-        if self.has_mesh_motion:
+        # initialize_mesh_motion if defined.
+        if self.use_mesh_motion_solver:
             self._mesh_moving_analysis.Initialize()
+            # since this is applied on the boundary surface
+            # we can fix MESH_DISPLACEMENT on all skin model part nodes.
+            nodes_array = ModelPartOperation(self.model, ModelPartOperation.OperationType.UNION, f"control_{self.GetName()}_skin", self.parameters["skin_model_part_names"].GetStringArray(), False).GetModelPart().Nodes
+            Kratos.VariableUtils().ApplyFixity(Kratos.MESH_DISPLACEMENT_X, True, nodes_array)
```
Comment (RezaNajian) [inline applications/OptimizationApplication/python_scripts/controls/shape/vertex_morphing_shape_control.py:88]: get an input to fix which directions, skin_model_part is not self explaining, I would suggest fixed model parts

File: applications/OptimizationApplication/python_scripts/controls/shape/vertex_morphing_shape_control.py (hunk 3 above)
Comment (RezaNajian) [reply applications/OptimizationApplication/python_scripts/controls/shape/vertex_morphing_shape_control.py:80]: Also do not use physical field as control field since it is very confusing and use the comments I made above that if the filter is explicit unfilter field returns zero otherwise implicit filter returns the control field

File: applications/OptimizationApplication/python_scripts/controls/shape/vertex_morphing_shape_control.py [hunk 5]
```diff
@@ -281,32 +127,44 @@ def MapGradient(self, physical_gradient_variable_container_expression_map: 'dict
...
         if not IsSameContainerExpression(new_control_field, self.GetEmptyField()):
-            raise RuntimeError(f"Updates for the required element container not found for control \"{self.GetName()}\". [ required model part name: {self.filter_model_part.FullName()}, given model part name: {new_control_field.GetModelPart().FullName()} ]")
+            raise RuntimeError(f"Updates for the required element container not found for control \"{self.GetName()}\". [ required model part name: {self.model_part.FullName()}, given model part name: {new_control_field.GetModelPart().FullName()} ]")
         if Kratos.Expression.Utils.NormL2(self.control_field - new_control_field) > 1e-15:
             # update the control SHAPE field
             control_update = new_control_field - self.control_field
             self.control_field = new_control_field
             # now update the physical field
             self._UpdateAndOutputFields(control_update)
             # update explicit filter
```
Comment (RezaNajian) [inline applications/OptimizationApplication/python_scripts/controls/shape/vertex_morphing_shape_control.py:146]: the update is called on any filter, please remove the explicit

File: applications/OptimizationApplication/python_scripts/controls/thickness/shell_thickness_control.py [hunk 6]
```diff
@@ -84,58 +36,48 @@ def __init__(self, name: str, model: Kratos.Model, parameters: Kratos.Parameters
...
-                "linear_solver_settings" : {}
-            },
-            "beta_value": 25.0,
-            "penalty_power": 1,
-            "output_all_fields": false,
-            "initial_physical_thickness":0.000001,
-            "physical_thicknesses": [],
-            "fixed_model_parts_and_thicknesses": {},
-            "utilities": []
+            "filter_settings"            : {},
```
Comment (RezaNajian) [inline applications/OptimizationApplication/python_scripts/controls/thickness/shell_thickness_control.py:39]: this PR is about the filters not changing the settings of the thickness control, please remove and put such changes into another separate PR. Again please do not change every thing without providing reasons and notices !

Comment (RezaNajian) [reply applications/OptimizationApplication/python_scripts/controls/thickness/shell_thickness_control.py:39]: beta settings have anything to do with update filters PR

File: applications/OptimizationApplication/python_scripts/controls/thickness/shell_thickness_control.py [hunk 7]
```diff
@@ -84,58 +36,48 @@ def __init__(self, name: str, model: Kratos.Model, parameters: Kratos.Parameters
...
-                "type" : "implicit",
-                "radius": 0.000000000001,
-                "linear_solver_settings" : {}
-            },
-            "beta_value": 25.0,
-            "penalty_power": 1,
-            "output_all_fields": false,
-            "initial_physical_thickness":0.000001,
-            "physical_thicknesses": [],
-            "fixed_model_parts_and_thicknesses": {},
```
Comment (RezaNajian) [inline applications/OptimizationApplication/python_scripts/controls/thickness/shell_thickness_control.py:97]: why such inputs and implementations are removed !!!
fixed_model_parts_and_thicknesses should be present

**Review by RezaNajian — CHANGES_REQUESTED** [review]
@sunethwarna thanks for the works, howevere there are many many changes and some of them like the comments in the thickness control and shape control have nothing to do with this PR which is about updating filters. As a general practice that is followed, please make separate PRs for separate topics, unfortunately I have to ask you to revert such changes and let's only focus on the filtering part.

File: applications/OptimizationApplication/custom_utilities/filtering/explicit_filter_utils.cpp (hunk 1 above)
Comment (sunethwarna) [reply applications/OptimizationApplication/custom_utilities/filtering/explicit_filter_utils.cpp:104]: Hi Reza, Now, the damping coefficients (D matrix) is computed somewhere else, and provided to the explicit filter. This is changed in this manner to allow different dampinh matrix computations to use the same filtering mechanicsm (at the moment, damping finds the neareset dampes node, and calculate the weights, this gives checkerboard pattern at the corners. So we have plans to implement new damping matrix computaion methods which does not give chekerboard patterns. For that this has to be changed).

I completely disagree with the performance issue. Because, in current and master implementations, both has search tree calls for filter neighbours and damping neareset neighbour search. Therefore, they are in par with the performance. Because, now D matrix is passed as an expression.

As I see, this is more generic, than being specific. In the case of filterig, explicit filters can have different damping matrices which is a nice feature I want to allow.

File: applications/OptimizationApplication/python_scripts/controls/shape/vertex_morphing_shape_control.py (hunk 2 above)
Comment (sunethwarna) [reply applications/OptimizationApplication/python_scripts/controls/shape/vertex_morphing_shape_control.py:41]: will be done.

File: applications/OptimizationApplication/python_scripts/controls/shape/vertex_morphing_shape_control.py (hunk 3 above)
Comment (sunethwarna) [reply applications/OptimizationApplication/python_scripts/controls/shape/vertex_morphing_shape_control.py:80]: understood, and will be changed

File: applications/OptimizationApplication/python_scripts/controls/shape/vertex_morphing_shape_control.py (hunk 4 above)
Comment (sunethwarna) [reply applications/OptimizationApplication/python_scripts/controls/shape/vertex_morphing_shape_control.py:88]: will be done :). Found a better modular way for this. I will put it.

File: applications/OptimizationApplication/python_scripts/controls/thickness/shell_thickness_control.py (hunk 7 above)
Comment (sunethwarna) [reply applications/OptimizationApplication/python_scripts/controls/thickness/shell_thickness_control.py:97]: I will change and put it those in seperate PR

**Comment by sunethwarna** [comment]
Closing this, as the issues/comments in this PR is addressed in #12201.

File: applications/OptimizationApplication/python_scripts/controls/shape/vertex_morphing_shape_control.py (hunk 5 above)
Comment (sunethwarna) [reply applications/OptimizationApplication/python_scripts/controls/shape/vertex_morphing_shape_control.py:146]: done :)
